Create new resource type: Microsoft.ResourceGraph/queries

ericc1103 opened the pull request:
### Latest improvements:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i><br>
### Contribution checklist:
- [x] I have reviewed the [documentation](https://github.com/Azure/azure-rest-api-specs#basics) for the workflow.
- [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.
- [x] The [OpenAPI Hub](https://aka.ms/openapiportal) was used for checking validation status and next steps.
### ARM API Review Checklist
- [x] Service team MUST add the "WaitForARMFeedback" label if the management plane API changes fall into one of the below categories. 
- adding/removing APIs.
- adding/removing properties.
- adding/removing API-version. 
- adding a new service in Azure.

Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs.
- [x] If you are blocked on ARM review and want to get the PR merged urgently, please get the ARM oncall for reviews (RP Manifest Approvers team under Azure Resource Manager service) from IcM and reach out to them. 
Please follow the link to find more details on [API review process](https://armwiki.azurewebsites.net/rp_onboarding/ResourceProviderOnboardingAPIRevieworkflow.html).

ericc1103 committed:
Adding Workbooks resource type.

ericc1103 committed:
Rename workbookDelete.json to WorkbookDelete.json

ericc1103 committed:
Fixed undefiend parameter issue with Delete verb.

ericc1103 committed:
Merge branch 'master' of https://github.com/ericc1103/azure-rest-api-specs

ericc1103 committed:
Update swagger to add subscriptionId

ericc1103 committed:
Fixed duplicated operation id issue

ericc1103 committed:
Removed links endpoint

ericc1103 committed:
Removed unused samples

ericc1103 committed:
Revert "Removed unused samples"

This reverts commit bed6bc30d4bc4bd5716fb9f907ac74f224b4b81d.

ericc1103 committed:
Revert "Removed links endpoint"

This reverts commit c245a85973ec62e036c09efc7375f77a5c4cf758.

ericc1103 committed:
Removed link logic and added get by resourceid

ericc1103 committed:
Applied PR comments

ericc1103 committed:
Added Resource Group parameter for GET. Removed get by sourceId since we are not using sourceId anymore

ericc1103 committed:
Merged with upstream. Modify Workbook Resource Type to include RG and remove unused endpoint

ericc1103 committed:
Resolved the merge conflicts with readme.md

ericc1103 committed:
Merge remote-tracking branch 'upstream/master'

ericc1103 committed:
Remove a location parameter since ARM routes it as we are using an tracted resource type

ericc1103 committed:
Created newer version (2018-04-01-preview) for workbook resource type.

ericc1103 committed:
Merge remote-tracking branch 'upstream/master'

ericc1103 committed:
Added Microsoft/Insights/Operations

ericc1103 committed:
Added Microsoft/Insights/Operations

ericc1103 committed:
Corrected the api version.

ericc1103 committed:
Made properties of PATCH operation as not required. Updated the version number to reflect the actual production one.

ericc1103 committed:
Updated Resource properties with the correct fields. Created new definition for PATCH operation since location is not required field

ericc1103 committed:
Update readme file

ericc1103 committed:
Uses plural for operation group name

ericc1103 committed:
Get the latest version from the upstream and resolved the merge conflict

ericc1103 committed:
Removed x-ms-mutability settings

ericc1103 committed:
Removed the location settings for PATCH

ericc1103 committed:
Resolved the merge conflicts

ericc1103 committed:
Added Microsoft.ResourceGraph/SavedQuery resource type

ericc1103 committed:
Fixed various validation errors

ericc1103 committed:
Updated ARG/resources to add a saved query support along with examples

ericc1103 committed:
Updated ARM/resources based on the feedback

ericc1103 committed:
Prepared for Swagger submittion

ericc1103 committed:
Merge remote-tracking branch 'upstream/master'

AutorestCI commented:
# Automation for azure-sdk-for-go

azuresdkci commented:
Can one of the admins verify this patch?

AutorestCI commented:
# Automation for azure-sdk-for-js

AutorestCI commented:
# Automation for azure-sdk-for-ruby

AutorestCI commented:
# Automation for azure-sdk-for-python

AutorestCI commented:
# Automation for azure-sdk-for-java

adxsdknet commented:
# Automation for azure-sdk-for-net 
A PR has been created for you: 
https://github.com/azure/azure-sdk-for-net/pull/5593 
.NET SDK Commits: 
https://github.com/adxsdknet/azure-sdk-for-net/pull/5593/commits/238769ca28e139d41c76d069fefd958694977f71

jhendrixMSFT reviewed (changes requested):
New swagger should have an entry in https://github.com/Azure/azure-rest-api-specs/blob/master/specification/resourcegraph/resource-manager/readme.md

ericc1103 commented:
@jhendrixMSFT Could you point me the documentation on how to map newly created files to readme?

jhendrixMSFT commented:
@ericc1103 it might not be documented, I'm following up on that.  Take a look at the other examples in the file, we can iterate on it in this PR.

ericc1103 committed:
Added saved query entry in readme.md

ericc1103 commented:
@jhendrixMSFT  Updated readme.md and it passes all tests.

chiragg4u commented:
@ericc1103 please review and get the approval internally in the team before publishing the swagger into the public REPO.

ericc1103 commented:
`please review and get the approval internally in the team before publishing the swagger into the public REPO.`

@chiragg4u What do you mean by that? How do I get an approval internally? From my past experiece, someone in ARM team contacted me to review it. Has the review process changed?

majastrz commented on specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2019-03-01-preview/savedquery.json:
```diff
@@ -0,0 +1,537 @@
+          }
+        ],
+        "responses": {
+          "201": {
```
>201 [](start = 11, length = 3)

As per [https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/resource-api-reference.md#delete-resource](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/resource-api-reference.md#delete-resource), this should be 200.

majastrz commented on specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2019-03-01-preview/savedquery.json:
```diff
@@ -0,0 +1,537 @@
+          "nextLinkName": null
+        }
+      }
+    }
```
This is redundant since resourcegraph.json already has it.

ericc1103 replied:
The resorucegraph.json has a lot more properties than it needs so we decided to use a slim version.

majastrz commented on specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2019-03-01-preview/savedquery.json:
```diff
@@ -0,0 +1,537 @@
+    }
+  },
+  "definitions": {
+    "Resource": {
```
>Resource [](start = 5, length = 8)

There are common definitions you can reuse via relative paths. Have you tried doing that?

ericc1103 replied:
This version has less properties than common one.

majastrz reviewed (changes requested):
Let's wait for the internal review first.

ericc1103 committed:
Changed HTTP status code of 201 to 200 for DELETE

ericc1103 committed:
Removed a name from payload since it is already specified in url

ericc1103 committed:
Removed location since it is a global resource and added eTag

ericc1103 committed:
Merge remote-tracking branch 'upstream/master'

ericc1103 committed:
Moved savedquery.json along with all examples into 2018-09-01-preview folder

ericc1103 committed:
Update saved query metadata in OperationsList.json

ericc1103 committed:
Updated readme.md file to address a version

ericc1103 committed:
Fixed a version for saved query

ericc1103 committed:
Fixed indentation of -input-file in readme

ericc1103 committed:
Renamed Saved => Graph

chiragg4u commented on specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/savedquery.json:
```diff
@@ -0,0 +1,536 @@
+          "type": "string",
+          "enum": [
+            "basic",
+            "extended"
```
Are we supporting extended to start with?

ericc1103 replied:
We are not supporting `extended` yet. Will remove it.

ericc1103 commented:
@chiragg4u I think I have updated Swagger based on your feedback (version, name, and etc) So would you please take a look at it again and remove `WaitForInternalTeamReview` tag? Please let me know if you want me to address other issues.

chiragg4u commented on specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/examples/OperationsList.json:
```diff
@@ -14,6 +14,15 @@
+            "display": {
+              "provider": "Microsoft Resource Graph",
+              "resource": "Queries",
+              "operation": "CRUD operations of saved queries",
```
Suggest removing reference to saved word. It's implicit as we are storing as the object. We can add description that it can be used to get the results instead of passing the query at run time.

chiragg4u commented on specification/resourcegraph/resource-manager/readme.md:
```diff
@@ -50,13 +49,15 @@ These settings apply only when `--tag=package-2019-04` is specified on the comma
 input-file:
-- Microsoft.ResourceGraph/preview/2018-09-01-preview/resourcegraph.json
+  - Microsoft.ResourceGraph/preview/2018-09-01-preview/resourcegraph.json
+  - Microsoft.ResourceGraph/preview/2018-09-01-preview/savedquery.json
```
Since we have a separate swagger for the queries; How will it be included in the SDK generations?

ericc1103 replied:
@chiragg4u Do you have any suggestion on this?

chiragg4u reviewed (commented):
Overall looks good.

ericc1103 commented on specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2019-03-01-preview/savedquery.json:
```diff
@@ -0,0 +1,537 @@
+{
```
I think we have to create new swagger file since this is new resource type.

majastrz replied:
It's not required to have separate swagger per resource type. It mostly lets you rev API versions separately per resource type. We're doing it this way in Blueprint.

---
In reply to: [276460279](https://github.com/Azure/azure-rest-api-specs/pull/5478#discussion_r276460279) [](ancestors = 276460279)

ericc1103 committed:
Merge remote-tracking branch 'upstream/master'

ericc1103 commented on specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2019-03-01-preview/examples/OperationsList.json:
```diff
@@ -0,0 +1,22 @@
+{
+  "parameters": {
+    "api-version": "2019-03-01-preview"
```
I think you are looking at the old file. It is already modified and fixed.

ericc1103 committed:
Applied PR comments

ericc1103 committed:
Restored extended enum value in ResultKind since we are using it

ericc1103 committed:
Removed the extended enum value again

ericc1103 commented:
> Let's wait for the internal review first.

@majastrz the internal review has been completed. Could we move forward?

majastrz commented:
> > Let's wait for the internal review first.
> 
> @majastrz the internal review has been completed. Could we move forward?

The ARM API on call should be able to help with that.

ericc1103 commented:
> > > Let's wait for the internal review first.
> > 
> > 
> > @majastrz the internal review has been completed. Could we move forward?
> 
> The ARM API on call should be able to help with that.

@majastrz could you please at least remove the tag you added? I think I have addressed your comments.

ericc1103 commented:
@KrisBash Could you please take a look at my change and kindly remove the tag if it meets your request?

chiragg4u commented on specification/resourcegraph/resource-manager/readme.md:
````diff
@@ -50,13 +49,15 @@ These settings apply only when `--tag=package-2019-04` is specified on the comma
 ``` yaml $(tag) == 'package-2018-09-preview'
 input-file:
-- Microsoft.ResourceGraph/preview/2018-09-01-preview/resourcegraph.json
+  - Microsoft.ResourceGraph/preview/2018-09-01-preview/resourcegraph.json
````
I'm not sure but should we refer to 2019-04-01 for this? Not sure how the SDK's work with different versions.

ericc1103 replied:
We have both 2019-04-01 and 2018-09-01 since we still need to support the old version.

chiragg4u reviewed (approved):
Approved.

majastrz commented on specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/graphquery.json:
```diff
@@ -0,0 +1,535 @@
+          }
+        },
+        "x-ms-pageable": {
+          "nextLinkName": null
```
>null [](start = 26, length = 4)

If you ever want to introduce paging in the future, you would have to introduce a new API version and you'll have constrain the number of results you return on the old API. My suggestion would be to set this to "nextLink" even if you are returning one page currently. That way you you can introduce paging later without breaking clients.

sanjaiganesh replied:
+1

---
In reply to: [278377348](https://github.com/Azure/azure-rest-api-specs/pull/5478#discussion_r278377348) [](ancestors = 278377348)

majastrz replied:
NextLink is still null here.

---
In reply to: [278412814](https://github.com/Azure/azure-rest-api-specs/pull/5478#discussion_r278412814) [](ancestors = 278412814,278377348)

ericc1103 replied:
Just updated as well.

majastrz replied:
See it now. Thanks!

---
In reply to: [279135228](https://github.com/Azure/azure-rest-api-specs/pull/5478#discussion_r279135228) [](ancestors = 279135228,278412814,278377348)

majastrz commented on specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/graphquery.json:
```diff
@@ -0,0 +1,535 @@
+        "query"
+      ],
+      "properties": {
+        "timeModified": {
```
>timeModified [](start = 9, length = 12)

Should this be a string or DateTime in the C# SDK? If it's the latter, you should add "format": "datetime"?

majastrz replied:
Btw in general, my suggestion would be to run autorest to generate one of the SDKs (c#, for example) and verify that what it generated seems reasonable.

---
In reply to: [278378779](https://github.com/Azure/azure-rest-api-specs/pull/5478#discussion_r278378779) [](ancestors = 278378779)

majastrz commented on specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/graphquery.json:
```diff
@@ -0,0 +1,535 @@
+          "default": "basic",
+          "type": "string",
+          "enum": [
+            "basic"
```
>basic [](start = 13, length = 5)

This enum doesn't have the "extended" value like the one below. Is that intentional?

ericc1103 replied:
We are not going to support Extended yet. It will come later when we enable that feature.

majastrz reviewed (changes requested):
Mostly looks ok. I added a couple of comments (ignore the old ones just look at the ones from 4/24 or newer).

sanjaiganesh commented on specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/graphquery.json:
```diff
@@ -0,0 +1,535 @@
+            "description": "The newly created graph query.",
+            "schema": {
+              "$ref": "#/definitions/GraphQueryResource"
+            }
```
Could you confirm whether PUT completes synchronously ? 
I see all operations are marked sync. please review and confirm

ericc1103 replied:
Yes it is

sanjaiganesh commented on specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/graphquery.json:
```diff
@@ -0,0 +1,535 @@
+          }
+        },
+        "x-ms-pageable": {
+          "nextLinkName": null
```
>"nextLinkName": null [](start = 10, length = 20)

same feedback here, to specify support for paging

sanjaiganesh commented on specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/graphquery.json:
```diff
@@ -0,0 +1,535 @@
+        }
+      }
+    },
+    "Operation": {
```
>Operation [](start = 5, length = 9)

"isDataAction" is missing

IsDataAction:  Required. Indicates whether the operation applies to data-plane. Set the value to "true" for data-plane operations and "false" for ARM/control-plane operations.

ericc1103 replied:
Not sure why we need this field. Could you elaborate more?

sanjaiganesh commented:
>         "value": [

As a note: all resource providers would need to include the "{Resource Provider Namespace}/register/action" operation in their response. This API is used to register for their service, and should include details about the operation (e.g. a localized name for the resource provider + any special considerations like PII release). Example values can be seen below:
Resource: "Storage Resource Provider"
Operation: "Registers the Storage Resource Provider"
Description: "Registers the subscription for the storage resource provider and enables the creation of storage accounts."

---
Refers to: specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/examples/OperationsList.json:8 in 2824f34. [](commit_id = 2824f349226d22bada62c59c9565a98fb242093f, deletion_comment = False)

ericc1103 commented:
> > ```
> >     "value": [
> > ```
> 
> As a note: all resource providers would need to include the "{Resource Provider Namespace}/register/action" operation in their response. This API is used to register for their service, and should include details about the operation (e.g. a localized name for the resource provider + any special considerations like PII release). Example values can be seen below:
> Resource: "Storage Resource Provider"
> Operation: "Registers the Storage Resource Provider"
> Description: "Registers the subscription for the storage resource provider and enables the creation of storage accounts."
> 
> Refers to: specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/examples/OperationsList.json:8 in [2824f34](https://github.com/Azure/azure-rest-api-specs/commit/2824f349226d22bada62c59c9565a98fb242093f). [](commit_id = [2824f34](https://github.com/Azure/azure-rest-api-specs/commit/2824f349226d22bada62c59c9565a98fb242093f), deletion_comment = False)

This resource doesn't require to be registered. We have marked it as `RegistrationFree`

ericc1103 committed:
Apply ARM API team comments

ericc1103 commented:
@sanjaiganesh & @majastrz, I have addressed the most of comments, please validate those and let me know.

ericc1103 committed:
Merge remote-tracking branch 'upstream/master'

ericc1103 committed:
Merge remote-tracking branch 'upstream/master'

ericc1103 committed:
Fix examples

ericc1103 committed:
Fix examples

ericc1103 committed:
Added nextLinkName

majastrz reviewed (approved):
Looks good.

ericc1103 commented:
@jhendrixMSFT, would you please remove the `requested changes` flag if everything looks OK?

ericc1103 commented:
@gauravkapila Thanks for the approval. But the `requested changes` should be removed that @jhendrixMSFT created. I have addressed his request so please remove it since you are DRI.

jhendrixMSFT committed:
add Go SDK build

jhendrixMSFT reviewed (approved):
Waiting for CI to finish, will merge assuming everything is good.

jhendrixMSFT committed:
make descriptions match

jhendrixMSFT commented on specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/graphquery.json:
```diff
@@ -0,0 +1,540 @@
+      },
+      "description": "Error message body that will indicate why the operation failed."
+    },
+    "OperationListResult": {
```
This is missing the `nextLink` field as defined in the `x-ms-pageable` extension for corresponding operation.

ericc1103 replied:
Thanks for catching this! Just updated.

ericc1103 committed:
Added nextLink definition in OperationListResult

ericc1103 committed:
Merge branch 'master' of https://github.com/ericc1103/azure-rest-api-specs

jhendrixMSFT commented on specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/graphquery.json:
```diff
@@ -0,0 +1,544 @@
+        }
+      }
+    },
+    "/providers/Microsoft.ResourceGraph/operations": {
```
This same path is defined in resourcegraph.json which is included in this same package, and since the return types are slightly different (nextLinkName is null in the other copy) it causes problems.  Can you please remove one of this?

ericc1103 replied:
@chiragg4u Do you want me to go ahead and fix this or want to handle it separately since it is not really a scope of my PR?

jhendrixMSFT replied:
It needs to be fixed, without it we cannot generate SDKs.

yangmsft replied:
@jhendrixMSFT Thanks for the heads-up. We will fix resourcegraph.json in a separate PR.

jhendrixMSFT replied:
Are you going to remove the duplicate entry from graphquery.json then?

jhendrixMSFT replied:
Sorry if I wasn't clear, the build for this new tag `package-2018-09-preview` is broken and must be fixed before we can merge this (see red CI).  You can either remove the duplicate entry added in graphquery.json or fix resourcegraph.json so the definitions are identical (I don't know which fix is correct so you will have to decide).

ericc1103 replied:
Just removed but seeing other issue and looking into it.

ericc1103 committed:
Removed a duplicated OperationListResult and reference to resourcegraph.json

ericc1103 committed:
Added nextLinkName: nextLink for x-ms-pageable in operations

jhendrixMSFT committed:
fix formatting and description

ericc1103 committed:
Removed extended from enum

ericc1103 committed:
Merge branch 'master' of https://github.com/ericc1103/azure-rest-api-specs
